File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/DP/Vacation.ts
function minCostTravel(days: number[], costs: number[]): number {
	const n = days.length;
	const max = days[n - 1];
	const dp: number[] = new Array(max + 1).fill(Infinity);

	let j = 0;
	dp[0] = 0; // Initialize the cost for day 0 as 0

	for (let i = 1; i <= max; i++) {
		if (days[j] === i) {
			j++;
			dp[i] = dp[i - 1] + costs[0];
			dp[i] = Math.min(dp[i], dp[Math.max(0, i - 7)] + costs[1]);
			dp[i] = Math.min(dp[i], dp[Math.max(0, i - 30)] + costs[2]);
		} else {
			dp[i] = dp[i - 1];
		}
	}

	return dp[max];
}

const days = [1, 3, 5, 7, 12, 20, 30];
const costs = [2, 7, 20];
console.log("Min cost is: " + minCostTravel(days, costs));

/*
 * Min cost is: 13
 */

File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/DP/Vacation.swift
import Foundation

// days are must travel days, costs are cost of tickets.
func minCostTravel(_ days : inout [Int], _ costs : inout [Int]) -> Int {
	let n : Int = days.count
	let mx : Int = days[n - 1]
	var dp : [Int] = Array(repeating: 0, count: mx + 1)
	var j : Int = 0
	var i : Int = 1
	while (i <= mx) {
		if (days[j] == i) { // That days is definitely travelled.
			j += 1
			dp[i] = dp[i - 1] + costs[0]
			dp[i] = min(dp[i], dp[max(0, i - 7)] + costs[1])
			dp[i] = min(dp[i], dp[max(0, i - 30)] + costs[2])
		} else {
			dp[i] = dp[i - 1]
		}
		i += 1
	}
	return dp[mx]
}

// Testing code.
var days : [Int] = [1, 3, 5, 7, 12, 20, 30]
var costs : [Int] = [2, 7, 20]
print("Min cost is:" + String(minCostTravel( &days, &costs)))

/*
Min cost is:13
*/